components/entitlement/org.wso2.carbon.identity.entitlement.endpoint/src/main/java/org/wso2/carbon/identity/entitlement/endpoint/auth/EntitlementAuthenticatorRegistry.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.entitlement.endpoint.auth;
+
+import org.apache.cxf.jaxrs.model.ClassResourceInfo;
+import org.apache.cxf.message.Message;
+
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * This stores the authenticators registered for Entitlement REST endpoints and returns the appropriate
+ * authenticator as requested by authentication filter associated with Entitlement REST endpoints.
+ */
+public class EntitlementAuthenticatorRegistry {
+
+    private static EntitlementAuthenticatorRegistry EntitlementAuthRegistry;
+    private static Map<Integer, EntitlementAuthenticationHandler> EntitlementAuthHandlers = new TreeMap<Integer,
+            EntitlementAuthenticationHandler>();
+
+    /**
+     * Initialize the EntitlementAuthenticatorRegistry Singleton
+     * Removed certain redundancies in singleton creation
+     *
+     * @return
+     */
+    public static EntitlementAuthenticatorRegistry getInstance() {
+        if (EntitlementAuthRegistry == null) {
+            synchronized (EntitlementAuthenticatorRegistry.class) {
+                if (EntitlementAuthRegistry == null) {
+                    EntitlementAuthRegistry = new EntitlementAuthenticatorRegistry();
+                    return EntitlementAuthRegistry;
+                }
+            }
+        }
```

Comment:
Double is locking for singleton class is known to be broken. http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html. Fix it properly.